File: leetcode/Q198_HouseRobber.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        length = len(nums)
        if length == 0:
            return 0
        elif length == 1:
            return nums[0]

        max_num = max(nums[0], nums[1])
        cache = [0, nums[0], nums[1]]
        for i in range(2, length):
            cache.append(max(nums[i] + cache[i-1], nums[i] + cache[i-2]))
            max_num = max(cache[i], max_num)
        return max(max_num, cache[-1])

File: src/algorithm1/dp/Q198_HouseRobber.java
package algorithm1.dp;

public class Q198_HouseRobber {
    class Solution {
        public int rob(int[] nums) {
            int length = nums.length;
            int[] cache = new int[length + 1];
            cache[1] = nums[0];
            for (int i = 2; i <= length; i++) {
                cache[i] = Math.max(cache[i-1], cache[i-2] + nums[i-1]);
            }
            return cache[length];
        }
    }
}
